crates/admin/src/schema_registry/updater.rs

```diff
@@ -0,0 +1,938 @@
+                    .components
+                    .get(component_name)
+                    .ok_or_else(|| {
+                        SchemaError::Subscription(SubscriptionError::SinkComponentNotFound(
+                            sink.clone(),
+                        ))
+                    })?;
+                if !component_schemas.handlers.contains_key(handler_name) {
+                    return Err(SchemaError::Subscription(
+                        SubscriptionError::SinkComponentNotFound(sink),
+                    ));
+                }
+
+                let ty = match component_schemas.ty {
+                    ComponentType::VirtualObject => EventReceiverComponentType::VirtualObject {
+                        ordering_key_is_key: false,
+                    },
+                    ComponentType::Service => EventReceiverComponentType::Service,
+                };
+
+                Sink::Component {
+                    name: component_name.to_owned(),
+                    handler: handler_name.to_owned(),
+                    ty,
+                }
+            }
+            _ => {
+                return Err(SchemaError::Subscription(
+                    SubscriptionError::InvalidSinkScheme(sink),
+                ))
+            }
+        };
+
+        let subscription = validator
+            .validate(Subscription::new(
+                id,
+                source,
+                sink,
+                metadata.unwrap_or_default(),
+            ))
+            .map_err(|e| SchemaError::Subscription(SubscriptionError::Validation(e.into())))?;
+
+        self.schema_information
+            .subscriptions
+            .insert(id, subscription);
+        self.modified = true;
+
+        Ok(id)
+    }
+
+    pub(crate) fn remove_subscription(&mut self, subscription_id: SubscriptionId) {
+        if self
+            .schema_information
+            .subscriptions
+            .remove(&subscription_id)
+            .is_some()
+        {
+            self.modified = true;
+        }
+    }
+
+    pub(crate) fn modify_component(&mut self, component_name: String, public: bool) {
+        if let Some(schemas) = self.schema_information.components.get_mut(&component_name) {
+            // Update the public field
+            if schemas.location.public != public {
+                schemas.location.public = public;
+                self.modified = true;
+            }
+
+            for h in schemas.handlers.values_mut() {
+                if h.target_meta.public != public {
+                    h.target_meta.public = public;
+                    self.modified = true;
+                }
+            }
+        }
+    }
+}
+
+fn check_reserved_name(name: &str) -> Result<(), ComponentError> {
+    if name.to_lowercase().starts_with("restate")
+        || name.to_lowercase().eq_ignore_ascii_case("openapi")
+    {
+        return Err(ComponentError::ReservedName(name.to_string()));
+    }
+    Ok(())
+}
```

Comment:
This is evidence that the component name type should be a NewType instead of &str.

Reply:
Yes, you are right. I will change it :-)